File: js/sandbox/simple_solver.js
// SimpleSolver for sandbox, debug.js, and tests.
//
// Provides a simplified interface to the solver that runs locally
// without web workers.

const { SudokuParser, toShortSolution } = await import('../sudoku_parser.js' + self.VERSION_PARAM);
const { SudokuConstraint } = await import('../sudoku_constraint.js' + self.VERSION_PARAM);
const { SudokuBuilder } = await import('../solver/sudoku_builder.js' + self.VERSION_PARAM);
const { GridShape } = await import('../grid_shape.js' + self.VERSION_PARAM);
const { Timer } = await import('../util.js' + self.VERSION_PARAM);
const { SolverStats } = await import('./solver_stats.js' + self.VERSION_PARAM);

/**
 * Simple synchronous solver interface.
 *
 * Each method takes constraints and runs a complete solve operation.
 * All operations run synchronously in the current thread.
 *
 * @example
 * const solver = new SimpleSolver();
 * for (const c of constraints) {
 *   const solution = solver.solution(c);
 *   console.log(solution?.toString());
 * }
 */
export class SimpleSolver {
  constructor() {
    this._state = null;
  }

  /**
   * Build solver from constraints and capture state after running.
   * @private
   */
  _build(constraints) {
    if (typeof constraints === 'string') {
      constraints = SudokuParser.parseText(constraints);
    } else if (Array.isArray(constraints)) {
      constraints = new SudokuConstraint.Container(constraints);
    }

    const timer = new Timer();
    let solver;
    timer.runTimed(() => {
      const resolved = SudokuBuilder.resolveConstraint(constraints);
      solver = SudokuBuilder.build(resolved);
    });

    const setupTimeMs = timer.elapsedMs();
    return {
      solver,
      shape: constraints.getShape(),
      captureState: () => {
        const state = solver.state?.();
        if (state) state.puzzleSetupTime = setupTimeMs;
        this._state = state;
      },
    };
  }

  /**
   * Find any solution.
   * @param {Object|Object[]|string} constraints
   * @returns {Solution|null}
   */
  solution(constraints) {
    const { solver, shape, captureState } = this._build(constraints);
    const values = solver.nthSolution(0);
    captureState();
    return values ? new Solution(values, shape) : null;
  }

  /**
   * Find the unique solution.
   * @param {Object|Object[]|string} constraints
   * @returns {Solution|null} The solution if exactly one exists, null otherwise
   */
  uniqueSolution(constraints) {
    const { solver, shape, captureState } = this._build(constraints);
    const first = solver.nthSolution(0);
    if (!first) {
      captureState();
      return null;
    }
    if (solver.nthSolution(1)) {
      captureState();
      return null;
    }
    captureState();
    return new Solution(first, shape);
  }

  /**
   * Iterator over solutions.
   * @param {Object|Object[]|string} constraints
   * @param {number} [limit]
   * @yields {Solution}
   */
  *solutions(constraints, limit) {
    const { solver, shape, captureState } = this._build(constraints);
    for (let n = 0; limit === undefined || n < limit; n++) {
      const values = solver.nthSolution(n);
      if (!values) break;
      captureState();
      yield new Solution(values, shape);
    }
    captureState();
  }

  /**
   * Count all solutions efficiently.
   * @param {Object|Object[]|string} constraints
   * @param {number} [limit] - Optional limit to count up to
   * @returns {number}
   */
  countSolutions(constraints, limit) {
    const { solver, captureState } = this._build(constraints);
    const count = solver.countSolutions(limit);
    captureState();
    return count;
  }

  /**
   * Find all true candidates (values appearing in valid solutions).
   * This is "All possibilities" mode in the UI.
   * @param {Object|Object[]|string} constraints
   * @param {number} [limit=1] - Candidate count limit to search up to
   * @returns {TrueCandidates|null}
   */
  trueCandidates(constraints, limit = 1) {
    const { solver, shape, captureState } = this._build(constraints);

    // Collect solutions via progress callback.
    const solutions = [];
    solver.setProgressCallback((extraState) => {
      if (extraState?.solutions) {
        solutions.push(...extraState.solutions);
      }
    });

    const counts = solver.solveAllPossibilities(limit);
    captureState();
    return counts ? new TrueCandidates(counts, shape, limit, solutions) : null;
  }

  /**
   * Validate that a layout (e.g., jigsaw) has valid solutions.
   * Non-layout constraints are ignored.
   * @param {Object|Object[]|string} constraints
   * @returns {Solution|null} A sample solution if valid, null if invalid
   */
  validateLayout(constraints) {
    const { solver, shape, captureState } = this._build(constraints);
    const values = solver.validateLayout();
    captureState();
    return values ? new Solution(values, shape) : null;
  }

  /**
   * Get statistics from the last solve operation.
   * @returns {SolverStats}
   */
  latestStats() {
    return new SolverStats(this._state);
  }
}

/**
 * Represents a single solution to a puzzle.
 */
export class Solution {
  constructor(values, shape) {
    this._values = values;
    this._shape = shape;
  }

  /**
   * Get the value at a cell.
   * @param {string|number} cellIdOrRow - Cell ID (e.g., 'R5C3') or row (1-indexed)
   * @param {number} [col] - Column (1-indexed), required if first arg is row
   * @returns {number} The value at the cell (1-9 for standard grid)
   */
  valueAt(cellIdOrRow, col) {
    return this._values[cellIndex(this._shape, cellIdOrRow, col)];
  }

  /**
   * Iterate over all cells in the solution.
   * @yields {{ cell: string, value: number }}
   */
  *[Symbol.iterator]() {
    for (let i = 0; i < this._values.length; i++) {
      yield {
        cell: this._shape.makeCellIdFromIndex(i),
        value: this._values[i],
      };
    }
  }

  /**
   * Get the solution as a short string (e.g., 81 chars for 9x9).
   * @returns {string}
   */
  toString() {
    return toShortSolution(this._values, this._shape);
  }

  /**
   * Compare with another solution or string.
   * @param {Solution|string} other
   * @returns {boolean}
   */
  equals(other) {
    const otherStr = other instanceof Solution ? other.toString() : other;
    return this.toString() === otherStr;
  }

  /**
   * Get the raw values array.
   * @returns {Uint8Array}
   */
  getArray() {
    return this._values;
  }
}

/**
 * Represents true candidates (values appearing in valid solutions).
 */
export class TrueCandidates {
  constructor(counts, shape, limit, solutions) {
    this._counts = counts;
    this._shape = shape;
    this._limit = limit;
    this._numValues = shape.numValues;
    this._solutions = solutions.map(s => new Solution(s, shape));
  }

  /**
   * Get all witness solutions.
   * @returns {Solution[]}
   */
  get witnessSolutions() {
    return this._solutions;
  }

  /**
   * Get all candidate values at a cell.
   * @param {string|number} cellIdOrRow - Cell ID (e.g., 'R5C3') or row (1-indexed)
   * @param {number} [col] - Column (1-indexed), required if first arg is row
   * @returns {number[]} Array of candidate values (1-indexed)
   */
  valuesAt(cellIdOrRow, col) {
    const idx = cellIndex(this._shape, cellIdOrRow, col);
    const baseIndex = idx * this._numValues;
    const values = [];
    for (let i = 0; i < this._numValues; i++) {
      if (this._counts[baseIndex + i] > 0) {
        values.push(i + 1);
      }
    }
    return values;
  }

  /**
   * Get the count for a specific value at a cell (capped to limit).
   * @param {string|number} cellIdOrRow - Cell ID (e.g., 'R5C3') or row (1-indexed)
   * @param {number} colOrValue - Column (1-indexed) if first arg is row, or value if first arg is cellId
   * @param {number} [value] - The value (1-indexed), required if first arg is row
   * @returns {number} Count of solutions containing this value
   */
  countAt(cellIdOrRow, colOrValue, value) {
    let idx, v;
    if (typeof cellIdOrRow === 'string') {
      idx = cellIndex(this._shape, cellIdOrRow);
      v = colOrValue;
    } else {
      idx = cellIndex(this._shape, cellIdOrRow, colOrValue);
      v = value;
    }
    const countIndex = idx * this._numValues + (v - 1);
    return Math.min(this._counts[countIndex], this._limit);
  }

  /**
   * Iterate over all non-zero candidates.
   * @yields {{ cell: string, value: number, count: number }}
   */
  *[Symbol.iterator]() {
    const numCells = this._shape.numCells;
    for (let i = 0; i < numCells; i++) {
      const baseIndex = i * this._numValues;
      for (let v = 0; v < this._numValues; v++) {
        const count = this._counts[baseIndex + v];
        if (count > 0) {
          yield {
            cell: this._shape.makeCellIdFromIndex(i),
            value: v + 1,
            count: Math.min(count, this._limit),
          };
        }
      }
    }
  }

  /**
   * Get candidates as a string (value for candidates, '.' for non-candidates).
   * @returns {string}
   */
  toString() {
    const numCells = this._shape.numCells;
    const baseCharCode = GridShape.baseCharCode(this._shape);
    const chars = [];
    for (let i = 0; i < numCells; i++) {
      const baseIndex = i * this._numValues;
      for (let v = 0; v < this._numValues; v++) {
        chars.push(this._counts[baseIndex + v] > 0
          ? String.fromCharCode(baseCharCode + v)
          : '.');
      }
    }
    return chars.join('');
  }
}

// Convert cell reference to cell index.
const cellIndex = (shape, cellIdOrRow, col) => {
  if (typeof cellIdOrRow === 'string') {
    return shape.parseCellId(cellIdOrRow).cell;
  }
  // Convert 1-indexed row/col to 0-indexed
  return shape.cellIndex(cellIdOrRow - 1, col - 1);
}

File: solver-wasm/src/simple_solver.rs
//! Simple synchronous solver interface — Rust port of `js/sandbox/simple_solver.js`.
//!
//! Provides a high-level API for solving puzzles without touching engine
//! internals. This is the API that anything other than the WASM entry
//! points should use (CLI, tests, etc.).
//!
//! # Examples
//!
//! ```
//! use solver_wasm::simple_solver::SimpleSolver;
//!
//! let mut solver = SimpleSolver::new();
//! let solution = solver.solution("53..7....6..195....98....6.8...6...34..8.3..17...2...6.6....28....419..5....8..79").unwrap();
//! assert!(solution.is_some());
//! ```

use std::fmt;
use std::time::Instant;

use crate::constraint::builder::SudokuBuilder;
use crate::constraint::parser;
use crate::grid_shape::GridShape;
use crate::solver::{Solver, SolverCounters};

// ---------------------------------------------------------------------------
// SimpleSolver
// ---------------------------------------------------------------------------

/// Simple synchronous solver interface.
///
/// Each method takes a constraint string, builds a solver, runs it, and
/// returns the result. Mirrors JS `SimpleSolver` in
/// `js/sandbox/simple_solver.js`.
///
/// The solver captures counters from the last operation, accessible via
/// [`latest_counters`](SimpleSolver::latest_counters).
pub struct SimpleSolver {
    counters: Option<SolverCounters>,
    /// Setup time (parse + build) in milliseconds, from the last operation.
    setup_time_ms: f64,
    /// Runtime (solver execution) in milliseconds, from the last operation.
    runtime_ms: f64,
}

impl SimpleSolver {
    pub fn new() -> Self {
        Self {
            counters: None,
            setup_time_ms: 0.0,
            runtime_ms: 0.0,
        }
    }

    /// Parse and build a solver from a constraint string, measuring setup time.
    ///
    /// Mirrors JS `SimpleSolver._build(constraints)`.
    fn build(input: &str) -> Result<(Solver, GridShape, f64), String> {
        let start = Instant::now();
        let parsed = parser::parse(input)?;
        let shape = parsed.shape;
        let solver = SudokuBuilder::build(&parsed)?;
        let setup_ms = start.elapsed().as_secs_f64() * 1000.0;
        Ok((solver, shape, setup_ms))
    }

    /// Find any solution.
    ///
    /// Mirrors JS `SimpleSolver.solution(constraints)`.
    pub fn solution(&mut self, input: &str) -> Result<Option<Solution>, String> {
        let (mut solver, shape, setup_ms) = Self::build(input)?;
        let start = Instant::now();
        let result = solver.nth_solution(0, &mut |_| {});
        self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;
        self.counters = Some(result.counters);
        self.setup_time_ms = setup_ms;
        Ok(result.solution.map(|values| Solution::new(values, shape)))
    }

    /// Find the unique solution. Returns `None` if there are zero or
    /// multiple solutions.
    ///
    /// Mirrors JS `SimpleSolver.uniqueSolution(constraints)`.
    pub fn unique_solution(&mut self, input: &str) -> Result<Option<Solution>, String> {
        let (mut solver, shape, setup_ms) = Self::build(input)?;
        let start = Instant::now();
        let first_result = solver.nth_solution(0, &mut |_| {});
        let first_values = match first_result.solution {
            Some(v) => v,
            None => {
                self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;
                self.counters = Some(first_result.counters);
                self.setup_time_ms = setup_ms;
                return Ok(None);
            }
        };
        let second_result = solver.nth_solution(1, &mut |_| {});
        self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;
        self.counters = Some(second_result.counters);
        self.setup_time_ms = setup_ms;
        if second_result.solution.is_some() {
            return Ok(None);
        }
        Ok(Some(Solution::new(first_values, shape)))
    }

    /// Collect up to `limit` solutions. If `limit` is `None`, collects
    /// all solutions (use with care — may be very large or infinite).
    ///
    /// Mirrors JS `SimpleSolver.solutions(constraints, limit)`.
    pub fn solutions(
        &mut self,
        input: &str,
        limit: Option<usize>,
    ) -> Result<Vec<Solution>, String> {
        let (mut solver, shape, setup_ms) = Self::build(input)?;
        let start = Instant::now();
        let mut solutions = Vec::new();
        let mut n = 0u64;
        loop {
            if let Some(lim) = limit {
                if solutions.len() >= lim {
                    break;
                }
            }
            let result = solver.nth_solution(n, &mut |_| {});
            self.counters = Some(result.counters);
            match result.solution {
                Some(values) => solutions.push(Solution::new(values, shape)),
                None => break,
            }
            n += 1;
        }
        self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;
        self.setup_time_ms = setup_ms;
        Ok(solutions)
    }

    /// Count solutions up to `limit` (0 or `None` = unlimited).
    ///
    /// Mirrors JS `SimpleSolver.countSolutions(constraints, limit)`.
    pub fn count_solutions(&mut self, input: &str, limit: Option<u64>) -> Result<u64, String> {
        let (mut solver, _shape, setup_ms) = Self::build(input)?;
        let start = Instant::now();
        let (count, counters) = solver.count_solutions(limit.unwrap_or(0), &mut |_| {});
        self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;
        self.counters = Some(counters);
        self.setup_time_ms = setup_ms;
        Ok(count)
    }

    /// Find all true candidates (values appearing in valid solutions).
    ///
    /// Mirrors JS `SimpleSolver.trueCandidates(constraints, limit)`.
    pub fn true_candidates(
        &mut self,
        input: &str,
        limit: u8,
    ) -> Result<Option<TrueCandidates>, String> {
        let (mut solver, shape, setup_ms) = Self::build(input)?;
        let start = Instant::now();
        let result = solver.solve_all_possibilities(limit, &mut |_| {});
        self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;

        if result.counters.solutions == 0 {
            self.counters = Some(result.counters);
            self.setup_time_ms = setup_ms;
            return Ok(None);
        }

        let witness_solutions: Vec<Solution> = result
            .solutions
            .into_iter()
            .map(|values| Solution::new(values, shape))
            .collect();

        self.counters = Some(result.counters);
        self.setup_time_ms = setup_ms;

        Ok(Some(TrueCandidates {
            counts: result.candidate_counts,
            shape,
            limit,
            num_values: shape.num_values as usize,
            witness_solutions,
        }))
    }

    /// Validate a layout (e.g. jigsaw) by checking whether any solution
    /// exists.
    ///
    /// Mirrors JS `SimpleSolver.validateLayout(constraints)`.
    pub fn validate_layout(&mut self, input: &str) -> Result<Option<Solution>, String> {
        let (mut solver, shape, setup_ms) = Self::build(input)?;
        let start = Instant::now();
        let result = solver.validate_layout(&mut |_| {});
        self.runtime_ms = start.elapsed().as_secs_f64() * 1000.0;
        self.counters = Some(result.counters);
        self.setup_time_ms = setup_ms;
        Ok(result.solution.map(|values| Solution::new(values, shape)))
    }

    /// Get counters from the last solve operation.
    ///
    /// Mirrors JS `SimpleSolver.latestStats()`.
    pub fn latest_counters(&self) -> Option<&SolverCounters> {
        self.counters.as_ref()
    }

    /// Setup time (parse + build) in milliseconds from the last operation.
    ///
    /// Mirrors JS `SolverStats.setupTimeMs` (from `state.puzzleSetupTime`).
    pub fn setup_time_ms(&self) -> f64 {
        self.setup_time_ms
    }

    /// Runtime (solver execution) in milliseconds from the last operation.
    ///
    /// Mirrors JS `SolverStats.runtimeMs` (from `state.timeMs`).
    pub fn runtime_ms(&self) -> f64 {
        self.runtime_ms
    }
}

// ---------------------------------------------------------------------------
// Solution
// ---------------------------------------------------------------------------

/// A single solution — 1-based values per cell.
///
/// Mirrors JS `Solution` in `js/sandbox/simple_solver.js`.
pub struct Solution {
    values: Vec<u8>,
    shape: GridShape,
}

impl Solution {
    fn new(values: Vec<u8>, shape: GridShape) -> Self {
        Self { values, shape }
    }

    /// Value at a cell index (0-based). Returns 0 for unsolved cells.
    pub fn value_at(&self, index: usize) -> u8 {
        self.values[index]
    }

    /// The raw values slice.
    pub fn values(&self) -> &[u8] {
        &self.values
    }

    /// The grid shape.
    pub fn shape(&self) -> GridShape {
        self.shape
    }
}

/// Display as a short solution string (e.g. 81 chars for 9×9).
///
/// Mirrors JS `Solution.toString()` → `toShortSolution()`.
impl fmt::Display for Solution {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = parser::to_short_solution(&self.values, self.shape);
        f.write_str(&s)
    }
}

impl PartialEq for Solution {
    fn eq(&self, other: &Self) -> bool {
        self.values == other.values && self.shape == other.shape
    }
}

impl PartialEq<str> for Solution {
    fn eq(&self, other: &str) -> bool {
        self.to_string() == other
    }
}

impl fmt::Debug for Solution {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Solution({})", self)
    }
}

// ---------------------------------------------------------------------------
// TrueCandidates
// ---------------------------------------------------------------------------

/// True candidates — per-cell-per-value counts from an exhaustive solve.
///
/// Mirrors JS `TrueCandidates` in `js/sandbox/simple_solver.js`.
pub struct TrueCandidates {
    /// Flat array: `counts[cell * num_values + (value - 1)]`.
    counts: Vec<u8>,
    shape: GridShape,
    limit: u8,
    num_values: usize,
    witness_solutions: Vec<Solution>,
}

impl TrueCandidates {
    /// All witness solutions found during the search.
    pub fn witness_solutions(&self) -> &[Solution] {
        &self.witness_solutions
    }

    /// Candidate values at a cell index (1-based values).
    pub fn values_at(&self, cell_index: usize) -> Vec<u8> {
        let base = cell_index * self.num_values;
        let mut values = Vec::new();
        for i in 0..self.num_values {
            if self.counts[base + i] > 0 {
                values.push((i + 1) as u8);
            }
        }
        values
    }

    /// Count for a specific value (1-based) at a cell, capped at `limit`.
    pub fn count_at(&self, cell_index: usize, value: u8) -> u8 {
        let idx = cell_index * self.num_values + (value as usize - 1);
        self.counts[idx].min(self.limit)
    }

    /// The grid shape.
    pub fn shape(&self) -> GridShape {
        self.shape
    }

    /// The raw candidate counts slice.
    pub fn counts(&self) -> &[u8] {
        &self.counts
    }
}

/// Display as a candidate string: value chars for candidates, '.' for
/// non-candidates.
///
/// Mirrors JS `TrueCandidates.toString()`.
impl fmt::Display for TrueCandidates {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let base_char = self.shape.base_char_code();
        let num_cells = self.shape.num_cells as usize;
        for i in 0..num_cells {
            let base_idx = i * self.num_values;
            for v in 0..self.num_values {
                if self.counts[base_idx + v] > 0 {
                    write!(f, "{}", char::from(base_char + v as u8))?;
                } else {
                    write!(f, ".")?;
                }
            }
        }
        Ok(())
    }
}

impl fmt::Debug for TrueCandidates {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "TrueCandidates({} cells, {} solutions)",
            self.shape.num_cells,
            self.witness_solutions.len()
        )
    }
}
